org-code-javabuilder/app/src/main/java/org/code/javabuilder/JavaRunner.java

```diff
@@ -50,40 +47,62 @@ public void compileAndRunUserProgram(UserProgram userProgram, Principal principa
-        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-            PrintStream out = new PrintStream(outputStream)) {
-          // set System.out to be a specific output stream in order to capture output of the
-          // program and send it back to the user
-          System.setOut(out);
-
-          this.compileRunService.sendMessages(principal.getName(), "Compiled!");
-          this.compileRunService.sendMessages(principal.getName(), "Running your program...");
-          runClass(tempFolder.toURI().toURL(), userProgram, principal);
-
-          // outputStream should now contain output of userProgram
-          outputStream.flush();
-          String result = outputStream.toString();
-          if (result.length() > 0) {
-            this.compileRunService.sendMessages(principal.getName(), result);
+    this.compileRunService.sendMessages(principal.getName(), "Compiling your program...");
+    boolean compileSuccess = compileProgram(userProgram, principal, tempFolder);
+
+    if (compileSuccess) {
+      this.compileRunService.sendMessages(principal.getName(), "Compiled!");
+      // We use a PipedInput and PipedOutput stream for I/O so we can control input to the user's
+      // program and output from the user's program.
+      // This try statement will close the streams automatically.
+      try (PipedInputStream inputStream = new PipedInputStream();
+          BufferedReader systemOutputReader =
+              new BufferedReader(new InputStreamReader(inputStream));
+          PipedOutputStream systemOutputStream = new PipedOutputStream(inputStream);
+          PrintStream out = new PrintStream(systemOutputStream)) {
+        // Set System.out to be a specific output stream in order to capture output of the
+        // program and send it back to the user
+        System.setOut(out);
+        this.systemInputStream = new PipedInputStream();
+        this.systemInputWriter = new PipedOutputStream(this.systemInputStream);
+        System.setIn(this.systemInputStream);
+
+        this.compileRunService.sendMessages(principal.getName(), "Running your program...");
+        JavaExecutorThread userRuntime =
+            new JavaExecutorThread(
+                tempFolder.toURI().toURL(), userProgram, principal, this.compileRunService);
+        userRuntime.start();
+
+        // TODO we'll need to stop execution after a certain time limit otherwise we'll hang here
+        // forever if a user writes infinite loops
+        while (userRuntime.isAlive() || systemOutputReader.ready()) {
+          if (systemOutputReader.ready()) {
+            // The user's program has produced output. Read it and pass it to the client console.
+            this.compileRunService.sendMessages(principal.getName(), systemOutputReader.readLine());
+          } else {
+            // The user's program is running, but there's no output. Let it continue running.
+            try {
+              Thread.sleep(200);
```

Comment:
Not sure if this is part of what will be refactored, but polling loops like this are generally not great for scalability.  Not sure what the best design here is off the top of my head, let me know if you want to brainstorm together.

Reply:
Some of this will be dependent on our AWS infrastructure, so I didn't want to sink too much time into it just yet. One possibility I've thinking about (curious for your thoughts): Rather than overriding System.in and System.out with a generic stream, override them with a thread-specific custom-written stream that logs directly to the user's channel. It would be set up similar to this: https://stackoverflow.com/a/15526391

Reply:
Interesting.  I'd need to spend more time with this to form an opinion but I think it's likely that we'll need to replace System.in and System.out with something custom.  I guess a general question is that, if System.out is shared across the process, how can we identify which student program output is coming from?  It sounds like we're thinking by-thread which is reasonable (but very strongly ties a build/run session with a thread).  (Also, it'd be good to double check on the thread safety guarantees for the various print calls to console.)